### ISSUE TITLE
"An error was encountered in a non-retryable streaming request."

### ISSUE THREAD

**Issue by jagthedrummer** [osiegmar/s3-publisher-action#78]
I've run into an issue where this action is having trouble uploading a file, but it doesn't offer many details about why.

The output of the upload section looks like this:

```
Run osiegmar/s3-publisher-action@v1
Sync files from electron-app/out/make/squirrel.windows/x64 (includes=**/*, excludes=, order=)
An error was encountered in a non-retryable streaming request.
Upload 1 new files
Uploading s3://***/Seshy-0.0.105 Setup.exe (type=application/x-msdos-program; Cache-Control=undefined)
Error: Error uploading to Seshy-0.0.105 Setup.exe
Upload 2 modified files (force=false)
Uploading s3://***/RELEASES (type=application/octet-stream; Cache-Control=undefined)
Uploading s3://***/Seshy-0.0.105-full.nupkg (type=application/octet-stream; Cache-Control=undefined)
Complete (1 added, 2 updated, 0 deleted)
```

In trying to debug the issue I added a step to list the contents of the directory that I'm trying to upload. The output of that section looks like this:

```
Run dir electron-app/out/make/squirrel.windows/x64
  
    Directory: D:\a\seshy\seshy\electron-app\out\make\squirrel.windows\x64
Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---           9/13/2024  5:43 PM             78 RELEASES
-a---           9/13/2024  5:43 PM      122187264 Seshy-0.0.105 Setup.exe
-a---           9/13/2024  5:43 PM      121427902 Seshy-0.0.105-full.nupkg
```

I'm using the action to upload build artifacts and it's working fine on a `macos-latest` action runner, but it's failing on a `windows-latest` runner. And even on `windows-latest` it's uploading 2 of the 3 files in question. It had been working fine on Windows until just this week, and now it fails consistently on that one file (`Seshy-0.0.105 Setup.exe`).

**Comment by osiegmar** [comment]
Currently, I have no idea what's happening. I've now added some more error handling/logging and also updated the dependencies. Could you try `osiegmar/s3-publisher-action@main` and share the output?

**Comment by jagthedrummer** [comment]
Thanks for looking into it, @osiegmar. Here's the output running agains `@main`.

```
Run osiegmar/s3-publisher-action@main
Sync files from electron-app/out/make/squirrel.windows/x64 (includes=**/*, excludes=, order=)
An error was encountered in a non-retryable streaming request.
Upload 1 new files
Uploading s3://***/Seshy-0.0.105 Setup.exe (type=application/x-msdos-program; Cache-Control=undefined)
Error: Error uploading to Seshy-0.0.105 Setup.exe: TimeoutError: socket hang up
Upload 2 modified files (force=false)
Uploading s3://***/RELEASES (type=application/octet-stream; Cache-Control=undefined)
Uploading s3://***/Seshy-0.0.105-full.nupkg (type=application/octet-stream; Cache-Control=undefined)
Complete (1 added, 2 updated, 0 deleted)
```

Looks like the underlying problem may be `TimeoutError: socket hang up`

**Comment by osiegmar** [comment]
That error is caused by the underlying AWS S3 library. I couldn't find much, but I guess multipart file upload could help for such large files. I will look into it...

**Comment by jagthedrummer** [comment]
Thanks, @osiegmar. I also wondered if a retry or timeout setting of some sort could help. The `.nupkg` file is only barely smaller than the `.exe` file, but may it just tiptoed across some threshold?

**Comment by osiegmar** [comment]
I read through the AWS documentation and code for a while.

Unfortunately the [AWS documentation](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-smithy-node-http-handler/Interface/NodeHttpHandlerOptions/) leaves `socketTimeout` undocumented.  In the source code, I could find this:

```typescript
req.setTimeout(this._socketTimeout || 3 * 60000, () => {
    if (socket) {
        socket.end();
    }
    handleResult(new Error(`Request timeout: ${info.options.path}`));
});
```

As the `socketTimeout` is not set, the default value of 3 minutes is used. The call to `socket.end()` seems to trigger the `Error: socket hang up` error, thrown by node.

Instead of increasing this timeout, I've implemented the multipart upload feature. Files larger than 16 MB are now uploaded in chunks of 16 MB – this is the same as the AWS CLI does. That should probably fix the issue.

Could you give `@main` a new try?

**Comment by jagthedrummer** [comment]
Thanks so much, @osiegmar! That did the trick!

This is the full output for the `s3-publisher-action` portion of the latest run:

```
Run osiegmar/s3-publisher-action@main
Sync files from electron-app/out/make/squirrel.windows/x64 (includes=**/*, excludes=, order=)
Upload 1 new files
Uploading s3://***/Seshy-0.0.105 Setup.exe (type=application/x-msdos-program; Cache-Control=undefined)
Finished uploading Seshy-0.0.105 Setup.exe
```

Maybe worth mentioning that previously there had been a few lines in the output about uploading "modified files" (even though they should have been identical files with a new timestamp) and that section of output was not there in my test run. The final `Complete` line with info about what all was done is also not there. Just wanted to mention it in case it's unintentional and/or in case it indicates some sort of silent error.

The "missing" lines that were there in a previous run:

```
Upload 2 modified files (force=false)
Uploading s3://***/RELEASES (type=application/octet-stream; Cache-Control=undefined)
Uploading s3://***/Seshy-0.0.105-full.nupkg (type=application/octet-stream; Cache-Control=undefined)
Complete (1 added, 2 updated, 0 deleted)
```

**Comment by jagthedrummer** [comment]
I deleted the previously-uploaded files and did a clean run to make sure that multiple files would be uploaded, and that worked as expected, but still no summary "Completed" line. Just for the record.

**Comment by osiegmar** [comment]
Good catch! I fixed this regression and just released v1.1.0 (also tagged by v1). Don't forget to switch back to that tag instead of `@main`.

Thanks for reporting this!

**Comment by jagthedrummer** [comment]
Everything is looking great here. Thanks again for the quick response on this, @osiegmar! 🚀 

```
Run osiegmar/s3-publisher-action@main
Sync files from electron-app/out/make (includes=**/*, excludes=, order=)
Upload 2 modified files (force=false)
Uploading s3://***/Seshy_development-0.0.105-arm64.dmg (type=application/x-apple-diskimage; Cache-Control=undefined)
Uploading s3://***/zip/darwin/arm64/Seshy_development-darwin-arm64-0.0.105.zip (type=application/zip; Cache-Control=undefined)
Finished uploading Seshy_development-0.0.105-arm64.dmg
Finished uploading zip/darwin/arm64/Seshy_development-darwin-arm64-0.0.105.zip
Complete (0 added, 2 updated, 0 deleted)
```
